submarine-workbench/workbench-server/src/test/java/org/apache/submarine/server/WorkbenchGitHubServerTest.java

```diff
@@ -0,0 +1,266 @@
+  private static final String LOCALPATH =
+      WorkbenchGitHubServerTest.class.getResource("").toString().substring(6) + REPO;
+
+  @After
+  public void restoreAllOperations() {
+    String token = System.getenv("gitToken");
+    if (token == null) {
+      LOG.warn("Token not set!");
+      return;
+    } else {
+      Map<String, String> map = new HashMap<>();
+      map.put("Authorization", "token " + token);
+      LOG.info("gitToken: {}", token);
+      sendHttpRequest("https://api.github.com/repos/" + OWNER + "/" + REPO + "/subscription",
+          map, null, "DELETE");
+      sendHttpRequest("https://api.github.com/user/starred/" + OWNER + "/" + REPO + "",
+          map, null, "DELETE");
+
+      String result = sendHttpRequest("https://api.github.com/repos/" + OWNER + "/" + REPO
+          + "/branches", map, null, "GET");
+      Gson gson = new Gson();
+      List<Map<String, String>> list = gson.fromJson(result, List.class);
+      for (Map<String, String> stringStringMap : list) {
+        if (BRANCHNAME.equals(stringStringMap.get("name"))) {
+          sendHttpRequest("https://api.github.com/repos/" + OWNER + "/" + REPO
+              + "/git/refs/heads/" + BRANCHNAME, map, null, "DELETE");
+        }
+      }
+      LOG.info("LOCALPATH: {}", LOCALPATH);
+      deleteDirectory(LOCALPATH);
+    }
+  }
+
+  public void deleteDirectory(String dir) {
+    File dirFile = new File(dir);
+    File[] files = dirFile.listFiles();
+    if (files != null) {
+      for (File file : files) {
+        if (file.isFile()) {
+          new File(file.getAbsolutePath()).delete();
+        } else if (file.isDirectory()) {
+          deleteDirectory(file.getAbsolutePath());
+        }
+      }
+      dirFile.delete();
+    }
+  }
+
+  @Test
+  public void addWatching() {
+    String token = System.getenv("gitToken");
+    if (token == null) {
+      LOG.warn("Token not set!");
+      return;
+    } else {
+      Map<String, String> map = new HashMap<>();
+      map.put("Authorization", "token " + token);
+      LOG.info("gitToken: {}", token);
+      sendHttpRequest("https://api.github.com/repos/" + OWNER + "/" + REPO + "/subscription",
+          map, null, "PUT");
+
+      String result = sendHttpRequest("https://api.github.com/repos/" + OWNER + "/" + REPO
+          + "/subscribers", null, null, "GET");
+      LOG.info("Watching result: {}", result);
+
+      Gson gson = new Gson();
+      List<Map<String, String>> list = gson.fromJson(result, List.class);
+      assertEquals(1, list.size());
+      String login = list.get(0).get("login");
+      LOG.info("login: {}", login);
+      assertEquals(OWNER, login);
+    }
+  }
+
+  @Test
+  public void deleteWatching() {
+    String token = System.getenv("gitToken");
+    if (token == null) {
+      LOG.warn("Token not set!");
+      return;
+    } else {
+      Map<String, String> map = new HashMap<>();
+      map.put("Authorization", "token " + token);
+      LOG.info("gitToken: {}", token);
```

Comment:
For security reasons, token cannot be printed.

Reply:
Done.